Prevent segfault on keyboard connection

Yaroslav-95 opened the pull request:
Sway was crashing on me everytime I connected or reconnected my Thinkpad usb keyboard. It turns out that for some reason, upon connection, the pointer to the keymap was NULL, therefore crashing Sway as it attempted to access the pointer in ipc_json_describe_input. I've added a check for that.

So far I've tested it and it doesn't crash on me anymore when connecting my keyboard. `swaymsg -t get_inputs` still shows the correct layouts for the keyboard, so I guess it sets the keymap pointer in a later event.

emersion commented:
Hm, this shouldn't happen. Do you have debug logs of the crash?

ErikReider commented:
This happened to me a few months ago but it was kinda random. I also think that it happened sometimes when connecting my Logitech G900 mouse. It hasn't happened in a while though...

Yaroslav-95 commented:
> Hm, this shouldn't happen. Do you have debug logs of the crash?

Sure. I've copied the part since I disconnect the keyboard. The crash happens when I connect back the keyboard.

```
00:12:40.076 [DEBUG] [wlr] [backend/libinput/events.c:208] Removing Lenovo ThinkPad Compact USB Keyboard with TrackPoint [6127:24647]
00:12:40.076 [DEBUG] [sway/input/input-manager.c:202] removing device: '6127:24647:Lenovo_ThinkPad_Compact_USB_Keyboard_with_TrackPoint'
00:12:40.076 [DEBUG] [sway/input/seat.c:950] removing device 6127:24647:Lenovo_ThinkPad_Compact_USB_Keyboard_with_TrackPoint from seat seat0
00:12:40.076 [DEBUG] [sway/input/keyboard.c:825] Removing keyboard 6127:24647:Lenovo_ThinkPad_Compact_USB_Keyboard_with_TrackPoint from group 0x55acfb0c7a70
00:12:40.076 [DEBUG] [sway/ipc-server.c:505] Sending input event
00:12:40.076 [DEBUG] [sway/ipc-server.c:958] Added IPC reply of type 0x80000015 to client 88 queue: { "change": "removed", "input": { "identifier": "6127:24647:Lenovo_ThinkPad_Compact_USB_Keyboard_with_TrackPoint", "name": "Lenovo ThinkPad Compact USB Keyboard with TrackPoint", "vendor": 6127, "product": 24647, "type": "keyboard", "xkb_layout_names": [ "English (US)", "Spanish", "Russian" ], "xkb_active_layout_index": 0, "xkb_active_layout_name": "English (US)", "libinput": { "send_events": "enabled" } } }
00:12:40.076 [DEBUG] [sway/ipc-server.c:958] Added IPC reply of type 0x80000015 to client 98 queue: { "change": "removed", "input": { "identifier": "6127:24647:Lenovo_ThinkPad_Compact_USB_Keyboard_with_TrackPoint", "name": "Lenovo ThinkPad Compact USB Keyboard with TrackPoint", "vendor": 6127, "product": 24647, "type": "keyboard", "xkb_layout_names": [ "English (US)", "Spanish", "Russian" ], "xkb_active_layout_index": 0, "xkb_active_layout_name": "English (US)", "libinput": { "send_events": "enabled" } } }
00:12:40.077 [DEBUG] [sway/ipc-server.c:535] Client 88 writable
00:12:40.077 [DEBUG] [sway/ipc-server.c:535] Client 98 writable
00:12:40.103 [DEBUG] [wlr] [backend/libinput/events.c:208] Removing Lenovo ThinkPad Compact USB Keyboard with TrackPoint [6127:24647]
00:12:40.103 [DEBUG] [sway/input/input-manager.c:202] removing device: '6127:24647:Lenovo_ThinkPad_Compact_USB_Keyboard_with_TrackPoint'
00:12:40.104 [DEBUG] [sway/input/seat.c:950] removing device 6127:24647:Lenovo_ThinkPad_Compact_USB_Keyboard_with_TrackPoint from seat seat0
00:12:40.104 [DEBUG] [sway/ipc-server.c:505] Sending input event
00:12:40.104 [DEBUG] [sway/ipc-server.c:958] Added IPC reply of type 0x80000015 to client 88 queue: { "change": "removed", "input": { "identifier": "6127:24647:Lenovo_ThinkPad_Compact_USB_Keyboard_with_TrackPoint", "name": "Lenovo ThinkPad Compact USB Keyboard with TrackPoint", "vendor": 6127, "product": 24647, "type": "pointer", "libinput": { "send_events": "enabled", "accel_speed": 1.0, "accel_profile": "flat", "natural_scroll": "disabled", "left_handed": "disabled", "middle_emulation": "disabled", "scroll_method": "on_button_down", "scroll_button": 274 } } }
00:12:40.104 [DEBUG] [sway/ipc-server.c:958] Added IPC reply of type 0x80000015 to client 98 queue: { "change": "removed", "input": { "identifier": "6127:24647:Lenovo_ThinkPad_Compact_USB_Keyboard_with_TrackPoint", "name": "Lenovo ThinkPad Compact USB Keyboard with TrackPoint", "vendor": 6127, "product": 24647, "type": "pointer", "libinput": { "send_events": "enabled", "accel_speed": 1.0, "accel_profile": "flat", "natural_scroll": "disabled", "left_handed": "disabled", "middle_emulation": "disabled", "scroll_method": "on_button_down", "scroll_button": 274 } } }
00:12:40.104 [DEBUG] [sway/input/input-manager.c:202] removing device: '6127:24647:Lenovo_ThinkPad_Compact_USB_Keyboard_with_TrackPoint'
00:12:40.104 [DEBUG] [sway/input/seat.c:950] removing device 6127:24647:Lenovo_ThinkPad_Compact_USB_Keyboard_with_TrackPoint from seat seat0
00:12:40.104 [DEBUG] [sway/input/keyboard.c:825] Removing keyboard 6127:24647:Lenovo_ThinkPad_Compact_USB_Keyboard_with_TrackPoint from group 0x55acfb0c7a70
00:12:40.104 [DEBUG] [sway/ipc-server.c:505] Sending input event
00:12:40.104 [DEBUG] [sway/ipc-server.c:958] Added IPC reply of type 0x80000015 to client 88 queue: { "change": "removed", "input": { "identifier": "6127:24647:Lenovo_ThinkPad_Compact_USB_Keyboard_with_TrackPoint", "name": "Lenovo ThinkPad Compact USB Keyboard with TrackPoint", "vendor": 6127, "product": 24647, "type": "keyboard", "xkb_layout_names": [ "English (US)", "Spanish", "Russian" ], "xkb_active_layout_index": 0, "xkb_active_layout_name": "English (US)", "libinput": { "send_events": "enabled", "accel_speed": 1.0, "accel_profile": "flat", "natural_scroll": "disabled", "left_handed": "disabled", "middle_emulation": "disabled", "scroll_method": "on_button_down", "scroll_button": 274 } } }
00:12:40.104 [DEBUG] [sway/ipc-server.c:958] Added IPC reply of type 0x80000015 to client 98 queue: { "change": "removed", "input": { "identifier": "6127:24647:Lenovo_ThinkPad_Compact_USB_Keyboard_with_TrackPoint", "name": "Lenovo ThinkPad Compact USB Keyboard with TrackPoint", "vendor": 6127, "product": 24647, "type": "keyboard", "xkb_layout_names": [ "English (US)", "Spanish", "Russian" ], "xkb_active_layout_index": 0, "xkb_active_layout_name": "English (US)", "libinput": { "send_events": "enabled", "accel_speed": 1.0, "accel_profile": "flat", "natural_scroll": "disabled", "left_handed": "disabled", "middle_emulation": "disabled", "scroll_method": "on_button_down", "scroll_button": 274 } } }
00:12:40.105 [DEBUG] [sway/ipc-server.c:535] Client 88 writable
00:12:40.105 [DEBUG] [sway/ipc-server.c:535] Client 98 writable
00:12:42.775 [DEBUG] [wlr] [backend/libinput/events.c:94] Added Lenovo ThinkPad Compact USB Keyboard with TrackPoint [6127:24647]
00:12:42.775 [DEBUG] [sway/input/input-manager.c:234] adding device: '6127:24647:Lenovo_ThinkPad_Compact_USB_Keyboard_with_TrackPoint'
00:12:42.775 [DEBUG] [sway/input/libinput.c:198] sway_input_configure_libinput_device('6127:24647:Lenovo_ThinkPad_Compact_USB_Keyboard_with_TrackPoint' on '6127:24647:Lenovo_ThinkPad_Compact_USB_Keyboard_with_TrackPoint')
00:12:42.775 [DEBUG] [sway/input/libinput.c:77] accel_set_speed(1.000000)
00:12:42.775 [DEBUG] [sway/input/libinput.c:88] accel_set_profile(1)
00:12:42.775 [DEBUG] [sway/ipc-server.c:505] Sending input event
00:12:41.581 [ERROR] [common/ipc-client.c:87] Unable to receive IPC response
Gdk-Message: 00:33:58.011: Error reading events from display: Broken pipe
Gdk-Message: 00:33:58.011: Error reading events from display: Broken pipe
00:12:41.659 [ERROR] [common/ipc-client.c:87] Unable to receive IPC response
Gdk-Message: 00:33:58.015: Error reading events from display: Connection reset by peer
```

Yaroslav-95 commented:
I also have a coredump. Here's the backtrace:

```
Core was generated by `build/sway/sway -d'.
Program terminated with signal SIGSEGV, Segmentation fault.
#0  0x00007f81771ae5a0 in xkb_keymap_num_layouts ()
   from /usr/lib/libxkbcommon.so.0
[Current thread is 1 (Thread 0x7f81765a2040 (LWP 15091))]
#0  0x00007f81771ae5a0 in xkb_keymap_num_layouts ()
    at /usr/lib/libxkbcommon.so.0
#1  0x00005567f4304540 in ipc_json_describe_input (device=0x5567f6e97230)
    at ../sway/ipc-json.c:987
        state = 0x0
        layouts_arr = 0x5567f6ed1bc0
        layout_idx = 21863
        keyboard = 0x5567f6e96ed0
        keymap = 0x0
        num_layouts = 21863
        __PRETTY_FUNCTION__ = "ipc_json_describe_input"
        object = 0x5567f6e553c0
#2  0x00005567f430700d in ipc_event_input
    (change=0x5567f4368d4a "libinput_config", device=0x5567f6e97230)
    at ../sway/ipc-server.c:509
        json = 0x5567f6dfc530
        json_string = 0x5567f6cfee60 "\340\212\312\366gU"
#3  0x00005567f4320193 in sway_input_configure_libinput_device
    (input_device=0x5567f6e97230) at ../sway/input/libinput.c:263
        ic = 0x5567f6c494e0
        device = 0x5567f6cfee60
        changed = true
#4  0x00005567f43171db in handle_new_input
    (listener=0x5567f6bf1a00, data=0x5567f6d008a0)
    at ../sway/input/input-manager.c:239
        input = 0x5567f6bf19c0
        device = 0x5567f6d008a0
        input_device = 0x5567f6e97230
        __PRETTY_FUNCTION__ = "handle_new_input"
        added = false
        seat = 0x5567f4385860 <sway_fmt>
#5  0x00007f81778f02c1 in wlr_signal_emit_safe
    (signal=0x5567f638d7a8, data=0x5567f6d008a0)
    at ../subprojects/wlroots/util/signal.c:29
        pos = 0x5567f6bf1a00
        l = 0x5567f6bf1a00
        cursor = 
          {link = {prev = 0x5567f6bf1a00, next = 0x7ffe76fe5bc0}, notify = 0x7f81778f020b <handle_noop>}
        end = 
          {link = {prev = 0x7ffe76fe5ba0, next = 0x5567f638d7a8}, notify = 0x7f81778f020b <handle_noop>}
#6  0x00007f81778a14e7 in new_input_reemit
    (listener=0x5567f638f3a0, data=0x5567f6d008a0)
    at ../subprojects/wlroots/backend/multi/backend.c:154
        state = 0x5567f638f390
#7  0x00007f81778f02c1 in wlr_signal_emit_safe
    (signal=0x5567f638f2c8, data=0x5567f6d008a0)
    at ../subprojects/wlroots/util/signal.c:29
        pos = 0x5567f638f3a0
        l = 0x5567f638f3a0
        cursor = 
          {link = {prev = 0x5567f638f3a0, next = 0x7ffe76fe5c60}, notify = 0x7f81778f020b <handle_noop>}
        end = 
          {link = {prev = 0x7ffe76fe5c40, next = 0x5567f638f2c8}, notify = 0x7f81778f020b <handle_noop>}
#8  0x00007f8177898bd0 in handle_device_added
    (backend=0x5567f638f2b0, libinput_dev=0x5567f6cfee60)
    at ../subprojects/wlroots/backend/libinput/events.c:109
        wlr_dev = 0x5567f6d008a0
        vendor = 6127
        product = 24647
        name = 0x5567f6e99c20 "Lenovo ThinkPad Compact USB Keyboard with TrackPoint"
        wlr_devices = 0x5567f6d00880
        dev = 0x5567f6f49de0
        tmp_dev = 0x5567f6f41d18
#9  0x00007f8177899343 in handle_libinput_event
    (backend=0x5567f638f2b0, event=0x5567f6e55430)
    at ../subprojects/wlroots/backend/libinput/events.c:242
        libinput_dev = 0x5567f6cfee60
        event_type = LIBINPUT_EVENT_DEVICE_ADDED
#10 0x00007f81778980d5 in handle_libinput_readable
    (fd=49, mask=1, _backend=0x5567f638f2b0)
    at ../subprojects/wlroots/backend/libinput/backend.c:58
        backend = 0x5567f638f2b0
        ret = 0
        event = 0x5567f6e55430
#11 0x00007f81771e407a in wl_event_loop_dispatch ()
    at /usr/lib/libwayland-server.so.0
#12 0x00007f81771e1be7 in wl_display_run () at /usr/lib/libwayland-server.so.0
#13 0x00005567f430a450 in server_run (server=0x5567f43855a0 <server>)
    at ../sway/server.c:261
#14 0x00005567f430986b in main (argc=2, argv=0x7ffe76fe6108)
    at ../sway/main.c:395
        verbose = 0
        debug = 1
        validate = 0
        allow_unsupported_gpu = 0
        long_options = 
            {{name = 0x5567f4367080 "help", has_arg = 0, flag = 0x0, val = 104}, {name = 0x5567f4367085 "config", has_arg = 1, flag = 0x0, val = 99}, {name = 0x5567f436708c "validate", has_arg = 0, flag = 0x0, val = 67}, {name = 0x5567f4367095 "debug", has_arg = 0, flag = 0x0, val = 100}, {name = 0x5567f436709b "version", has_arg = 0, flag = 0x0, val = 118}, {name = 0x5567f43670a3 "verbose", has_arg = 0, flag = 0x0, val = 86}, {name = 0x5567f43670ab "get-socketpath", has_arg = 0, flag = 0x0, val = 112}, {name = 0x5567f43670ba "unsupported-gpu", has_arg = 0, flag = 0x0, val = 117}, {name = 0x5567f43670ca "my-next-gpu-wont-be-nvidia", has_arg = 0, flag = 0x0, val = 117}, {name = 0x0, has_arg = 0, flag = 0x0, val = 0}}
        config_path = 0x0
        usage = 0x5567f4366af0 "Usage: sway [options] [command]\n\n  -h, --help", ' ' <repeats 13 times>, "Show help message and quit.\n  -c, --config <config>  Specify a config file.\n  -C, --validate         Check the validity of the config file, th"...
        c = -1
```

emersion commented:
Hm, it seems like we send a `libinput_config` IPC event before the `added` event…

shinanca commented:
Confirmed. With this patch no more segfault.
Arch sway-git.

Yaroslav-95 commented:
I've updated the patch to improve the fix as per https://github.com/swaywm/sway/issues/6873#issuecomment-1285289179_

emersion reviewed (approved):
LGTM, thanks!
